gtda/utils/validation.py

```diff
@@ -77,102 +76,121 @@ def check_graph(X):
 
 # Check the type and range of numerical parameters
 def validate_params(parameters, references):
+    """Function to automate the validation of hyperparameters.
+
+    Parameters
+    ----------
+    parameters : dict, required
+        Dictionary in which the keys are hyperparameter names and the
+        corresponding values are hyperparameter values.
+
+    references : dict, required
+        Dictionary in which the keys are hyperparameter names and the
+        corresponding values are lists. The first element of that list is a
+        type and the second is either ``None``, a tuple containing two
+        elements representing the bounds of the range of values the
+        hyperparameters can take, or a list containing all possible allowed
+        values for the hyperparameter.
+
+    """
     for key in references.keys():
+        # Check type
         if not isinstance(parameters[key], references[key][0]):
-            raise TypeError("Parameter {} is of type {}"
-                            " while it should be of type {}"
-                            "".format(key, type(parameters[key]),
-                                      references[key][0]))
-        if len(references[key]) == 1:
-            continue
-        if references[key][0] == list or \
+            raise TypeError(
+                f"Parameter {key} is of type {type(parameters[key])} while it "
+                f"should be of type {references[key][0]}.")
+
+        # If the key is a list, tuple, or numpy array, check element by element
+        if references[key][0] == list or references[key][0] == tuple or \
            references[key][0] == np.ndarray:
             for parameter in parameters[key]:
+                # If an element has to be an int, it can be passed as a float
+                # that has no decimals, else we check the type of the element
                 if references[key][1][0] == int:
```

Comment:
There seem to be some important but tacit assumptions about the form of the list `references[key][1]` here and in what follows. Could you clarify the intent?